src/shared/messaging/test/port-rpc-test.js

```diff
@@ -175,23 +175,158 @@ describe('PortRPC', () => {
+    const sender = new PortRPC();
+    const receiver = new PortRPC();
+    const closeHandler = sinon.stub();
 
-      assert.isNull(rpc.disconnect());
+    receiver.on('close', closeHandler);
+    receiver.connect(port2);
+    sender.connect(port1);
+    await waitForMessageDelivery();
 
-      const { port1 } = new MessageChannel();
-      rpc.connect(port1);
+    closeHandler.resetHistory();
+    window.dispatchEvent(new Event('unload'));
+    await waitForMessageDelivery();
 
-      assert.equal(rpc.disconnect(), port1);
-      assert.isNull(rpc.disconnect());
+    assert.calledWith(closeHandler);
+  });
 
-      rpc.call('test');
+  it('should only invoke "close" handler once', async () => {
+    const { port1, port2 } = new MessageChannel();
+    const sender = new PortRPC();
+    const receiver = new PortRPC();
+    const closeHandler = sinon.stub();
+
+    receiver.on('close', closeHandler);
+    receiver.connect(port2);
+    sender.connect(port1);
+
+    // Invoke "close" manually. In a real app it will be invoked when the
+    // window is unloaded and/or the PortRPC is destroyed.
+    sender.call('close');
+    sender.call('close');
+
+    await waitForMessageDelivery();
+
+    assert.calledOnce(closeHandler);
+  });
+
+  /** Transfer a MessagePort to another frame and return the transferred port. */
+  async function transferPort(port, targetWindow) {
+    const transferredPort = new Promise(resolve => {
+      targetWindow.addEventListener('message', e => {
+        if (e.ports[0]) {
+          resolve(e.ports[0]);
+        }
+      });
+    });
+    targetWindow.postMessage({}, '*', [port]);
+    return transferredPort;
+  }
+
+  describe('Safari <= 15 workaround', () => {
+    const safariUserAgent =
+      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.2 Safari/605.1.15';
+
+    let removeWorkaround;
+    let childFrame;
+
+    beforeEach(() => {
+      childFrame = document.createElement('iframe');
+      document.body.append(childFrame);
+
+      removeWorkaround = installPortCloseWorkaroundForSafari(safariUserAgent);
+    });
+
+    afterEach(() => {
+      removeWorkaround();
+      childFrame.remove();
+    });
+
+    it('transfers port to parent frame and sends "close" event from there when window is unloaded', async () => {
+      const { port1, port2 } = new MessageChannel();
+      const transferredPort = await transferPort(
+        port1,
+        childFrame.contentWindow
+      );
+
+      const sender = new PortRPC({
+        userAgent: safariUserAgent,
+        currentWindow: childFrame.contentWindow,
+      });
+      const receiver = new PortRPC();
+      const closeHandler = sinon.stub();
+
+      receiver.on('close', closeHandler);
+      receiver.connect(port2);
+      sender.connect(transferredPort);
       await waitForMessageDelivery();
-      assert.notCalled(testMethod);
+
+      closeHandler.resetHistory();
```

Comment:
Doesn't this send the message that `closeHandler` could be triggered several times?

Reply:
I changed the various `resetHistory` calls to asserts.